consensus/safety-rules/src/safety_rules_manager.rs

```diff
@@ -0,0 +1,86 @@
+// Copyright (c) The Libra Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{
+    local_client::LocalClient, persistent_storage::PersistentStorage, InMemoryStorage,
+    OnDiskStorage, SafetyRules, TSafetyRules,
+};
+use consensus_types::common::Payload;
+use libra_config::config::{NodeConfig, SafetyRulesBackend, SafetyRulesConfig};
+use libra_types::crypto_proxies::ValidatorSigner;
+use std::sync::{Arc, RwLock};
+
+pub struct SafetyRulesManagerConfig {
+    storage: Option<Box<dyn PersistentStorage>>,
+    validator_signer: Option<ValidatorSigner>,
+}
+
+impl SafetyRulesManagerConfig {
+    pub fn new(config: &mut NodeConfig) -> Self {
+        let private_key = config
+            .consensus
+            .consensus_keypair
+            .take_private()
+            .expect("Failed to take Consensus private key, key absent or already read");
+
+        let author = config
+            .validator_network
+            .as_ref()
+            .expect("Missing validator network")
+            .peer_id;
+
+        let validator_signer = ValidatorSigner::new(author, private_key);
+        Self::new_with_signer(validator_signer, &config.consensus.safety_rules)
+    }
+
+    pub fn new_with_signer(validator_signer: ValidatorSigner, config: &SafetyRulesConfig) -> Self {
+        let storage = match &config.backend {
+            SafetyRulesBackend::InMemoryStorage => InMemoryStorage::default_storage(),
+            SafetyRulesBackend::OnDiskStorage(config) => {
+                if config.default {
+                    OnDiskStorage::default_storage(config.path().clone())
+                        .expect("Unable to allocate SafetyRules storage")
+                } else {
+                    OnDiskStorage::new_storage(config.path().clone())
+                        .expect("Unable to instantiate SafetyRules storage")
+                }
+            }
+        };
+
+        Self {
+            storage: Some(storage),
+            validator_signer: Some(validator_signer),
+        }
+    }
+}
+
+enum SafetyRulesWrapper<T> {
+    Local(Arc<RwLock<SafetyRules<T>>>),
+}
+
+pub struct SafetyRulesManager<T> {
```

Comment:
My original intent was to have a single layer here, but the problem is that ChainedBftProvider knows about a specific T (Vec<Transaction>) and unless I want to create a boxed trait it seems I cannot use a typed manager directly. Also for InMemoryStorage the "service" needs to be kept consistent or we need to handle passing around state (yuck). This also allows us to remove the Arc around ValidatorSigner which seems to clean up the SafetyRules so that it doesn't incur some of the tech debt created by the current path between consensus provider through epoch manager. @dmitri-perelman and I began a discussion about simplifying this flow at which point we should just have a manager and no need for the config above. I plan to append to the existing issues so that we don't skip over this point.

Reply:
I probably agree with @zekun000 here.
What would happen if we'd not have `SafetyRulesWrapper` and `LocalClient` structs? 
Could `SafetyRulesManager` generate and return `Arc<RwLock<SafetyRulesT>>`?

Reply:
This hides the details of what is behind the Boxed TSafetyRules. Arc<RwLock< ... is not a Boxed safety rules. The implementation just needs a boxed TSafetyRules and we can hide all the logic for handling whether it is remote or local within that boxed object.

Longer term, we won't have both Config and Manager, we would just have one or the other but due to the trait relationships and encapsulation within the code, it seems necessary or we have to expose Arc<RwLock<Box<TSafetyRules>>> to EpochManager.

Reply:
Why do you need Boxed TSafetyRules? You're saying `Arc<RwLock<SafetyRulesT>>` would not work?